Mathlib/CategoryTheory/Sites/SheafHom.lean

```diff
@@ -0,0 +1,204 @@
+/-
+Copyright (c) 2023 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+
+import Mathlib.CategoryTheory.Sites.Over
+
+/-! Internal hom of sheaves
+
+In this file, given two sheaves `F` and `G` on a site `(C, J)` with values
+in a category `A`, we shall define a sheaf of types
+`Sheaf.internalHom F G` which sends `X : C` to the type of morphisms
+between the restrictions of `F` and `G` to the categories `Over X`.
+
+We first define `Presheaf.internalHom F G` when `F` and `G` are
+presheaves `Cᵒᵖ ⥤ A` and show that it is a sheaf when `G` is a sheaf.
+
+-/
+
+universe v v' u u'
+
+namespace CategoryTheory
+
+open Category Opposite Limits
+
+variable {C : Type u} [Category.{v} C] {J : GrothendieckTopology C}
+  {A : Type u'} [Category.{v'} A]
+
+namespace Presheaf
+
+variable (F G : Cᵒᵖ ⥤ A)
+
+/-- Given two presheaves `F` and `G` on a category `C` with values in a category `A`,
+this `internalHom F G` is the presheaf of types which sends an object `X : C`
+to the type of morphisms between the "restrictions" of `F` and `G` to the category `Over X`. -/
+@[simps! obj]
+def internalHom : Cᵒᵖ ⥤ Type _ where
```

Comment:
`internalHom` sounds a bit strange as it's not valued in the original category `A`. `yoneda` feels more appropriate, especially if you show it's functorial in F.

Reply:
It should not be too hard to deduce an internal hom in `Sheaf J A` when `A` has a limit-preserving, isomorphism-reflecting functor to types though. I'm not sure it's interesting in other cases. 
Internal also refers to the fact that it's internal to sheaf categories.

Reply:
A itself need to have internal homs, right? Like the category of abelian groups or modules over a CommRing.

Reply:
Right, of course!

Reply:
I am not convinced of `yoneda` as a name, but would `presheafHom` (and `sheafHom` for the sheaf version) be ok? (I do not intend to develop the `A`-valued version soon.)

Reply:
Sounds good to me

Reply:
I have fixed the name as `presheafHom`.